File: lib/beam_bot/exchanges/domain/trading_pair.ex
defmodule BeamBot.Exchanges.Domain.TradingPair do
  @moduledoc """
  This module is responsible for managing the trading pairs.
  """

  use Ecto.Schema
  import Ecto.Changeset
  alias BeamBot.Exchanges.Domain.Exchange

  @type t :: %__MODULE__{
          id: integer() | nil,
          exchange_id: integer() | nil,
          exchange: Exchange.t() | nil,
          symbol: String.t() | nil,
          base_asset: String.t() | nil,
          quote_asset: String.t() | nil,
          min_price: Decimal.t() | nil,
          max_price: Decimal.t() | nil,
          tick_size: Decimal.t() | nil,
          min_qty: Decimal.t() | nil,
          max_qty: Decimal.t() | nil,
          step_size: Decimal.t() | nil,
          min_notional: Decimal.t() | nil,
          is_active: boolean(),
          status: String.t(),
          inserted_at: DateTime.t() | nil,
          updated_at: DateTime.t() | nil
        }

  @derive {Jason.Encoder,
           only: [
             :symbol,
             :base_asset,
             :quote_asset,
             :min_price,
             :max_price,
             :tick_size,
             :min_qty,
             :max_qty,
             :step_size,
             :min_notional,
             :is_active,
             :status
           ]}
  schema "trading_pairs" do
    belongs_to :exchange, Exchange
    field :symbol, :string
    field :base_asset, :string
    field :quote_asset, :string
    field :status, :string
    field :min_price, :decimal
    field :max_price, :decimal
    field :tick_size, :decimal
    field :min_qty, :decimal
    field :max_qty, :decimal
    field :step_size, :decimal
    field :min_notional, :decimal
    field :is_active, :boolean, default: true

    timestamps()
  end

  @doc false
  def changeset(trading_pair, attrs) do
    trading_pair
    |> cast(attrs, [
      :exchange_id,
      :symbol,
      :base_asset,
      :quote_asset,
      :min_price,
      :max_price,
      :tick_size,
      :min_qty,
      :max_qty,
      :step_size,
      :min_notional,
      :is_active,
      :status
    ])
    |> validate_required([
      :exchange_id,
      :symbol,
      :base_asset,
      :quote_asset,
      :status
    ])
    |> foreign_key_constraint(:exchange_id)
  end
end

File: rust/src/domain/trading_pairs/trading_pair.rs
use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use std::fmt;

#[derive(Debug, Clone)]
pub struct TradingPair {
    pub id: Option<i64>,
    pub symbol: String,
    pub base_asset: String,
    pub quote_asset: String,
    pub min_price: Option<Decimal>,
    pub max_price: Option<Decimal>,
    pub tick_size: Option<Decimal>,
    pub min_qty: Option<Decimal>,
    pub max_qty: Option<Decimal>,
    pub step_size: Option<Decimal>,
    pub min_notional: Option<Decimal>,
    pub is_active: bool,
    pub status: String,
    pub is_margin_trading: bool,
    pub is_spot_trading: bool,
    pub exchange_id: i64,
    pub sync_start_time: Option<DateTime<Utc>>,
    pub sync_end_time: Option<DateTime<Utc>>,
}

impl fmt::Display for TradingPair {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.symbol)
    }
}

impl TradingPair {
    pub fn new(
        symbol: String,
        base_asset: String,
        quote_asset: String,
        status: String,
        is_margin_trading: bool,
        is_spot_trading: bool,
        exchange_id: i64,
        sync_start_time: Option<DateTime<Utc>>,
        sync_end_time: Option<DateTime<Utc>>,
        min_price: Option<Decimal>,
        max_price: Option<Decimal>,
        tick_size: Option<Decimal>,
        min_qty: Option<Decimal>,
        max_qty: Option<Decimal>,
        step_size: Option<Decimal>,
        min_notional: Option<Decimal>,
    ) -> Self {
        Self {
            id: None,
            symbol,
            base_asset,
            quote_asset,
            min_price,
            max_price,
            tick_size,
            min_qty,
            max_qty,
            step_size,
            min_notional,
            is_active: true,
            status,
            is_margin_trading,
            is_spot_trading,
            exchange_id,
            sync_start_time,
            sync_end_time,
        }
    }
}
